lib/Runtime/ByteCode/ByteCodeEmitter.cpp

```diff
@@ -3113,60 +3127,63 @@ void ByteCodeGenerator::EmitOneFunction(ParseNode *pnode)
             EnsureImportBindingScopeSlots(pnode, funcInfo);
         }
 
-        // Emit all scope-wide function definitions before emitting function bodies
-        // so that calls may reference functions they precede lexically.
-        // Note, global eval scope is a fake local scope and is handled as if it were
-        // a lexical block instead of a true global scope, so do not define the functions
-        // here. They will be defined during BeginEmitBlock.
-        if (!(funcInfo->IsGlobalFunction() && this->IsEvalWithNoParentScopeInfo()))
-        {
-            DefineFunctions(funcInfo);
-        }
-        DefineUserVars(funcInfo);
-
         ::BeginEmitBlock(pnode->sxFnc.pnodeScopes, this, funcInfo);
 
-        if (paramScope != nullptr && !paramScope->GetCanMergeWithBodyScope())
-        {
-            // If the param scope has own scope then we should move the params only after the inner scope slot is created
-            HomeArguments(funcInfo);
-        }
-
-        if (pnode->sxFnc.pnodeBodyScope != nullptr)
-        {
-            ::BeginEmitBlock(pnode->sxFnc.pnodeBodyScope, this, funcInfo);
-        }
         DefineLabels(funcInfo);
 
         if (pnode->sxFnc.HasNonSimpleParameterList())
         {
             Scope* bodyScope = funcInfo->GetBodyScope();
 
+            this->InitBlockScopedNonTemps(funcInfo->root->sxFnc.pnodeScopes, funcInfo);
+
             if (!paramScope->GetCanMergeWithBodyScope())
             {
-                // Right now the scope stack is like this: outer scope -> body scope -> param scope (Pushed during BeginEmitBlock)
-                // When param and body scopes are not merged they need to be considered at the same level. So we cannot have body
-                // scope as the enclosing scope of param scope. Both body scope and the param scope should have the outer scope as
-                // the enclosing scope. When emitting the formals the stack should be like this: outer scope -> param scope -> formals.
-                // When emitting the body the stack should be like this: outer scope -> body scope -> body statements.
-                // For that pop the param scope, pop the body scope and then push the param scope again.
-                PopScope();
-                Assert(this->GetCurrentScope()->GetScopeType() == ScopeType_FunctionBody);
+                byteCodeFunction->SetParamAndBodyScopeNotMerged();
+
+                HomeArguments(funcInfo);
+
+                // Pop the body scope before emitting the default args
```

Comment:
Okay. I remember we discussed restructuring the StartEmitFunction level. Is there a reason why we can't do that? It would be much better, IMO, to keep the scope stack logically consistent throughout.